dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java

```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package org.apache.cxf.microprofile.client.cdi;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import javax.interceptor.InvocationContext;
+
+
+class MPRestClientInvocationContextImpl implements InvocationContext {
+
+    private final Object target;
+
+    private final Method method;
+
+    private Object[] args;
+
+    private int index = 0;
+
+    private final List<InterceptorInvoker> interceptorInvokers;
+
+    private final Map<String, Object> contextData = new HashMap<>();
+
+    private final Callable<Object> callable;
+    /**
+     * @param target
+     * @param method
+     * @param args
+     * @param interceptorInvokers
+     */
+    MPRestClientInvocationContextImpl(Object target, Method method, Object[] args, List<InterceptorInvoker> interceptorInvokers, Callable<Object> callable) {
+        this.target = target;
+        this.method = method;
+        this.args = args == null ? new Object[] {} : args;
+        this.interceptorInvokers = interceptorInvokers;
+        this.callable = callable;
+    }
+
+    boolean hasNextInterceptor() {
+        return index < interceptorInvokers.size();
+    }
+
+    protected Object invokeNextInterceptor() throws Exception {
+        int oldIndex = index;
+        try {
+            return interceptorInvokers.get(index++).invoke(this);
```

Comment:
This will go hilariously wrong if there are interceptors after the fault tolerance interceptor.
* `Retry` can cause later interceptors to be called more than once
* `Asynchronous` can cause later interceptors to be run on a different thread, possibly after all the earlier interceptors have completed.

Reply:
I can add a comment to that effect here - but I think the FT interceptor should always be the last one invoked due to the ordering imposed in CDIInterceptorWrapperImpl.